frontend/src/main/java/starter/project/frontend/CarController.java

```diff
@@ -1,36 +1,56 @@
 package starter.project.frontend;
 
-import org.springframework.web.bind.annotation.PostMapping;
-import org.springframework.web.bind.annotation.RequestBody;
-import org.springframework.web.bind.annotation.ResponseBody;
-import org.springframework.web.bind.annotation.RestController;
-import starter.project.domain.Car;
 
-import java.util.ArrayList;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
+import org.springframework.http.HttpStatus;
+import org.springframework.stereotype.Component;
+import org.springframework.web.bind.annotation.*;
+import org.springframework.web.server.ResponseStatusException;
+import starter.project.domain.Car;
+import java.util.*;
 
 @RestController
 public class CarController {
+
     private List<Car> carNewList = new ArrayList<>();
+    Map<String, Car> map = new HashMap<>();
+
+    @PostMapping(path = "/car", consumes = {"application/json"})
+    public List<Car> addCar(@RequestBody Car car) {
+        carNewList.add(car);
+        return carNewList;
+    }
+
+    @GetMapping(path = "/car/{carId}")
+    public Car getCar(@PathVariable("carId") int carId) {
+
+        for (Car car : carNewList) {
+            if (car.getId() == (carId)) {
+                return car;
+            }
+        }
+        throw new ResponseStatusException(
+                HttpStatus.NOT_FOUND, "entity not found");
+    }
+
+    @GetMapping(path = "/cars/{list}")
+    public List<Car> getBatchCars(@PathVariable List list) {
 
+        List<Car> newList = new ArrayList<>();
+        for (Map.Entry<String, Car> m : map.entrySet()) {
```

Comment:
This is kind of inefficient. You'd expect the map of all keys in a database to be waaaaaaaay bigger than the list of ids being asked for. So it's iterating a map (which is usually a bad idea) that is possibly orders of magnitude bigger. 

It should iterate the list and do map.get (which is O(1) time) so the whole thing becomes O(N) where N is the list size.

Reply:
resolved